### DISCUSSION TITLE
Add type hints to MulensModel

### DISCUSSION THREAD

**Discussion by ketozhang** [rpoleski/MulensModel#46] [Ideas]
Type hints (or annotations) are great self-documenting code. Along with a type checker, you can have the benefit of statically typed languages in your editor.

To introduce type hints to MulensModel there are a few compatibility code to add. For Python 3.6 ([it's currently end of life](https://endoflife.date/python)), we'll need to add the `typing_extensions` package. For Python 3.7+, we need to add `from __futures__ import annotations` to use any future typing features introduced in the latest version of Python.

## Use Case

In the code below, `MulensData` can be instantiated with data passed into `data_list` which accepts a list of three collections (by the docstring some users may infer you can pass in anything that can become an `ndarray`).

```py
class MulensData

  def __init__(data_list=None, ...):
    ...

```

Instead of relying on the users to correctly interpret docstring, we may add type hinting to explictly let the type checking check that `data_list` contains 3 items and each item can be converted to ndarray via:

```py
from numpy.typing import ArrayLike 
from typing import Optional
class MulensData

  def __init__(data_list: Optional[tuple[ArrayLike, ArrayLike, ArrayLike]]=None, ...):
    ...
```

For code reusage, we may make use of Type Alias:

```py
from numpy.typing import ArrayLike   # numpy >= 1.20
from typing import Optional

LightCurveArray = tuple[ArrayLike, ArrayLike, ArrayLike]]

class MulensData

  def __init__(data_list: Optional[LightCurveArray]=None, ...):
    ...
```

Although I much prefer using dictionaries

```py
from numpy.typing import ArrayLike 
from typing import Optional, TypedDict

class LightCurveDict(TypedDict):
  time: ArrayLike
  mag: ArrayLike
  mag_err: ArrayLike
  

class MulensData

  def __init__(data_list: Optional[LightCurveDict]=None, ...):
    ...
```

**Comment by ketozhang** [comment]
The suggestion isn't to add type hints to every part of the code (since type hints are optional and has no effect on runtime).
Rather I suggest we add them to public interfaces that users would touch and start using them internally for newer code.

**Reply by rpoleski** [reply]
This seems to be quite a lot of work to add type hints to all public functions. I won't do it myself, but I'm open to contributions implemented by others.

One question - is sphinx producing documentation properly for type hints?

**Reply by ketozhang** [reply]
Yes, I am willing to make contributions. 

`TODO` We will need to identify popular interfaces then create issue ticket for each interface this. 

> One question - is sphinx producing documentation properly for type hints?

Yes. You may use this https://github.com/tox-dev/sphinx-autodoc-typehints
Flask is an example of usage: https://flask.palletsprojects.com/en/2.2.x/api/
